crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,413 @@
+    ) -> SuiResult {
+        // Mismatch between node epoch and checkpoint epoch after startup
+        // crash recovery is invalid
+        let checkpoint_epoch = checkpoint.epoch();
+        let node_epoch = self.authority_state.epoch();
+        assert_eq!(
+            checkpoint_epoch, node_epoch,
+            "Epoch mismatch after startup recovery. checkpoint epoch: {:?}, node epoch: {:?}",
+            checkpoint_epoch, node_epoch,
+        );
+
+        let next_committee = checkpoint.summary().next_epoch_committee.clone();
+
+        if next_committee.is_some() {
+            self.end_of_epoch = true;
+        }
+
+        let highest_scheduled = checkpoint.sequence_number();
+        let state = self.authority_state.clone();
+        let store = self.checkpoint_store.clone();
+        let metrics = self.metrics.clone();
+
+        pending.push_back(spawn_monitored_task!(async move {
+            while let Err(err) =
+                execute_checkpoint(checkpoint.clone(), state.clone(), store.clone()).await
+            {
+                error!(
+                    "Error while executing checkpoint, will retry in 1s: {:?}",
+                    err
+                );
+                tokio::time::sleep(Duration::from_secs(1)).await;
+                metrics.checkpoint_exec_errors.inc();
+            }
+
+            (checkpoint, next_committee)
+        }));
+
+        self.highest_scheduled_seq_num = Some(highest_scheduled);
+
+        Ok(())
+    }
+}
+
+pub async fn execute_checkpoint(
+    checkpoint: VerifiedCheckpoint,
+    authority_state: Arc<AuthorityState>,
+    checkpoint_store: Arc<CheckpointStore>,
+) -> SuiResult<Vec<SignedTransactionEffects>> {
+    let txes = checkpoint_store
+        .get_checkpoint_contents(&checkpoint.content_digest())?
+        .unwrap_or_else(|| {
+            panic!(
+                "Checkpoint contents for digest {:?} does not exist",
+                checkpoint.content_digest()
+            )
+        })
+        .into_inner();
+
+    execute_transactions(txes, authority_state).await
+}
+
+async fn execute_transactions(
+    transactions: Vec<ExecutionDigests>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<SignedTransactionEffects>> {
+    let tx_digests: Vec<TransactionDigest> = transactions.iter().map(|tx| tx.transaction).collect();
+    let txns: Vec<VerifiedCertificate> = authority_state
+        .database
+        .epoch_store()
+        .multi_get_pending_certificate(&tx_digests)?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    // TODO once https://github.com/MystenLabs/sui/pull/6157 is landed,
+    // replace with call to `authority_state.add_pending_certs_and_effects`,
+    // which handles reading effects for shared object version numbers
+    // and enqueing transactions with TransactionManager.
+    let digest_to_effects: HashMap<TransactionDigest, TransactionEffects> = authority_state
+        .database
+        .get_effects(&tx_digests)?
```

Comment:
I don't understand this part. Did we write effects from checkpoints directly into the `executed_effects` table before their execution? If we can already read effects for `txns`, the execution becomes a no-op.

Reply:
agreed - this method reads from executed_effects, but we want to read from the `effects` which is populated by data sync, right?

Reply:
Good catch, thanks